### PR TITLE
Strip leading `.` from cookie domain names.

### PR DESCRIPTION
**Pull request by icemac**
If the value of the `domain` attribute in the cookie starts with `.` Zope currently breaks.
But according to https://www.rfc-editor.org/rfc/rfc6265#section-4.1.2.3 it should be allowed.

I added a test and a fix, without the fix the new test breaks with:

```
Error in test test_domain (ZPublisher.tests.test_cookie.CookieParameterRegistryTests)
Traceback (most recent call last):
  File "/.../lib/python3.7/unittest/case.py", line 59, in testPartExecutor
    yield
  File "/.../lib/python3.7/unittest/case.py", line 628, in run
    testMethod()
  File "/.../Zope/src/ZPublisher/tests/test_cookie.py", line 150, in test_domain
    _, v = convertCookieParameter("domain", ".zope.dev")
  File "/.../Zope/src/ZPublisher/cookie.py", line 152, in convert
    return nn, (value if value is None else self[nn](value))
  File "/.../Zope/src/ZPublisher/cookie.py", line 249, in domain_converter
    value = ".".join(to_str(ToASCII(nameprep(c))) for c in u_value.split("."))
  File "/.../Zope/src/ZPublisher/cookie.py", line 249, in <genexpr>
    value = ".".join(to_str(ToASCII(nameprep(c))) for c in u_value.split("."))
  File "/.../lib/python3.7/encodings/idna.py", line 73, in ToASCII
    raise UnicodeError("label empty or too long")
UnicodeError: label empty or too long
```

### CODE REVIEW

**Review by dataflake — APPROVED** [review]

**Review by d-maurer — APPROVED** [review]

File: src/ZPublisher/cookie.py [hunk 1]
```diff
@@ -242,6 +242,10 @@ def domain_converter(value):
         return value
     from encodings.idna import ToASCII
     from encodings.idna import nameprep
```
Comment (icemac) [inline src/ZPublisher/cookie.py:244]: @d-maurer Is it necessary to do these imports here instead of at the top of the module?

Comment (dataflake) [reply src/ZPublisher/cookie.py:244]: I don't see a good reason for that, just move them to the top.

File: src/ZPublisher/cookie.py [hunk 2]
```diff
@@ -242,7 +242,14 @@ def domain_converter(value):
...
-    return ".".join(to_str(ToASCII(nameprep(c))) for c in u_value.split("."))
+    add_leading_dot = False
+    if u_value.startswith("."):
+        u_value = u_value[1:]
+        add_leading_dot = True
+    value = ".".join(to_str(ToASCII(nameprep(c))) for c in u_value.split("."))
+    if add_leading_dot:
+        value = f".{value}"
```
[the remark this answers is not in the document]
Comment (icemac) [reply src/ZPublisher/cookie.py]: Fixed in https://github.com/zopefoundation/Zope/pull/1041/commits/39617784c47bf1695cca6d164a500ab789af0599.

**Comment by d-maurer** [comment]
Michael Howitz wrote at 2022-5-17 23:25 -0700:
***@***.*** commented on this pull request.
>
>
>
>> @@ -242,6 +242,10 @@ def domain_converter(value):
>         return value
>     from encodings.idna import ToASCII
>     from encodings.idna import nameprep
>
***@***.*** Is it necessary to do these imports here instead of at the top of the module?

They are here for documentation/modularity purposes:
this documents that they are special to this converter and not used
elsewhere.


Maybe, we should consider
"https://github.com/zopefoundation/Zope/issues/1033":
meanwhile, I believe that using the IDNA encoding (in this place) is wrong
altogether and that it should get replaced by normal URL quoting
(with safe characters defined by
"https://datatracker.ietf.org/doc/html/rfc3986.html#section-3.2.2").
IDNA is likely necessary only in interactions with the DNS;
Browsers apparently do not expect IDNA encoding in the `host` part
of an URL but normal URL quoting.

**Comment by dataflake** [comment]
I'm going to merge this, #1033 is a separate issue for a separate PR.

**Pull request merged**